ENH: Add the ability to block callback signals

greglucas opened the pull request:
## PR Summary

CallbackRegistry objects gain a context manager `.block()` method that can be used to temporarily block callback signals from being processed.


```python
with callbacks.block():
    callbacks.process("test")  # Does nothing
with callbacks.block(signal="test"):
    callbacks.process("test")  # Does nothing
```

Closes #20802 
Would help with disabling norm update signals in #19553

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

story645 commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -280,6 +280,31 @@ def process(self, s, *args, **kwargs):
+
+        Parameters
+        ----------
+        signal : str, optional
+            The callback signal to block, default is None to block all signals
```
```suggestion
        signal : str, default: None
            The callback signal to block, default blocks all signals
```
and this may need an almost doctest style example string? I didn't get how to use this w/o looking at the tests.  Maybe go long form-'signal_name' ?

greglucas replied:
`signal` is used in the process/disconnect functions as well, so I think that we should stick to that here.

I added a quick example of how to use this to the `CallbackRegistry()` docstring. Is that what you were looking for, or something else with the doctest style example?

I added this under the API change note, I am not sure if this is truly a great "next what's new" feature because I didn't expect many external users to use it, but I could move it over if you think that would be a better place for it and add some of the test examples there.

story645 replied:
I think it'd be a useful short thing to mention in what's new. Plenty of folks build interactive tools & I think the temporary blocking could be useful for 'em.

story645 reviewed (commented):
I think this could be really useful for debugging interactive stuff, so tentatively 👍

story645 commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -280,6 +286,31 @@ def process(self, s, *args, **kwargs):
+
+        Parameters
+        ----------
+        signal : str, default: None
+            The callback signal to block, default blocks all signals
```
```suggestion
        signal : str, default: None
            The callback signal to block, default is to block all signals
```
yes correcting myself here, but I'm not sure which is the correct grammatical construction.

greglucas replied:
I put it back to optional, which is apparently preferred if the value isn't used explicitly as a value I think...? https://numpydoc.readthedocs.io/en/latest/format.html#parameters

I also made it into two sentences which may be easier to follow now. No worries on the nits if you have other suggestions. Better to get it right the first time.

story645 commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -143,6 +144,11 @@ class CallbackRegistry:
         >>> callbacks.disconnect(id_eat)
```
Since you're touching this, mind putting  a new line before `.disconnect` is introduced?

story645 reviewed (approved):
minor nits, but liking the new docs :)

timhoffm commented:
What do you think about naming this `blocked()`? I like to use context manager names that are not imperative verbs and cannot be confused with simple functions. There is no general convention on this in the community, but I think `with callbacks.blocked()` reads quite nicely.

greglucas commented:
@timhoffm, apparently there is a preferred convention and your suggestion lines up with that :)
https://www.python.org/dev/peps/pep-0343/#examples
I changed to `blocked` and moved the change note over to "next what's new" and included a short example there.

story645 commented:
So is maybe annoying, but I think this needs a short API changes note too. 

Also, and I might be confused about the functionality here, but would it be possible for the what's new example to be slight less toy/use like one of the auto registrated callbacks or something?

greglucas commented:
I came up with another toy-like example, but with interactivity on the figure canvas and disabling key-press events. Did you have another example/area in mind to apply this? It looks like the widgets keep their callbacks private in `_observers`, so I guess this doesn't really help the end user currently over there.

story645 commented:
Using fig.canvas.callbacks is perfect - I was unsure whether the functionality extended to the fig callbacks and this clears it right up & I think would be a common use case 😄
